spring-cloud-gcp-autoconfigure/src/main/java/com/google/cloud/spring/autoconfigure/pubsub/PubSubBeanProcessor.java

```diff
@@ -0,0 +1,144 @@
+			String subscriptionName = subscription.getKey();
+			PubSubConfiguration.Subscriber selectiveSubscriber = subscriberMap.get(subscriptionName);
+			Integer selectiveExecutorThreads = selectiveSubscriber.getExecutorThreads();
+			if (selectiveExecutorThreads != null) {
+				String threadName = "gcp-pubsub-subscriber-" + subscriptionName;
+				String beanName = "threadPoolScheduler_" + subscriptionName;
+				ThreadPoolTaskScheduler selectiveScheduler = createAndRegisterSchedulerBean(selectiveExecutorThreads,
+						threadName, beanName, beanDefinitionRegistry);
+				String fullyQualifiedName = PubSubSubscriptionUtils
+						.toProjectSubscriptionName(subscriptionName, projectIdProvider.getProjectId()).toString();
+				this.threadPoolTaskSchedulerMap.putIfAbsent(fullyQualifiedName, selectiveScheduler);
+			}
+		}
+
+		// Register global threadPoolTaskScheduler configuration bean
+		PubSubConfiguration.Subscriber globalSubscriber = pubSubProperties.getSubscriber();
+		Integer globalExecutorThreads = globalSubscriber.getExecutorThreads();
+		Integer numThreads = globalExecutorThreads != null ? globalExecutorThreads
+				: PubSubConfiguration.DEFAULT_EXECUTOR_THREADS;
+		this.globalScheduler = createAndRegisterSchedulerBean(numThreads, "global-gcp-pubsub-subscriber",
+				"globalThreadPoolScheduler", beanDefinitionRegistry);
+
+	}
+
+	@Override
+	public void postProcessBeanFactory(ConfigurableListableBeanFactory configurableListableBeanFactory)
+			throws BeansException {
+		// Do nothing.
+	}
+
+	/**
+	 * Creates a {@link ThreadPoolTaskScheduler} and registers it as a bean.
+	 * @param executorThreads number of executor threads
+	 * @param threadName thread name
+	 * @param beanName bean name
+	 * @param beanDefinitionRegistry bean registry
+	 * @return a {@link ThreadPoolTaskScheduler}
+	 */
+	private ThreadPoolTaskScheduler createAndRegisterSchedulerBean(Integer executorThreads, String threadName,
+			String beanName,
+			BeanDefinitionRegistry beanDefinitionRegistry) {
+		ThreadPoolTaskScheduler scheduler = createThreadPoolTaskScheduler(executorThreads, threadName);
+		beanDefinitionRegistry.registerBeanDefinition(beanName,
+				BeanDefinitionBuilder.genericBeanDefinition(ThreadPoolTaskScheduler.class, () -> scheduler)
+						.getBeanDefinition());
+		return scheduler;
+	}
+
+	/**
+	 * Creates {@link ThreadPoolTaskScheduler} given the number of executor threads and a
+	 * thread name.
+	 * @param executorThreads number of executor threads
+	 * @param threadName thread name prefix to set for the scheduler
+	 * @return thread pool scheduler
+	 */
+	ThreadPoolTaskScheduler createThreadPoolTaskScheduler(Integer executorThreads, String threadName) {
+		ThreadPoolTaskScheduler scheduler = new ThreadPoolTaskScheduler();
+		scheduler.setPoolSize(executorThreads);
+		scheduler.setThreadNamePrefix(threadName);
+		scheduler.setDaemon(true);
+		return scheduler;
+	}
+
+	public ConcurrentMap<String, ThreadPoolTaskScheduler> getThreadPoolSchedulerMap() {
```

Comment:
It's in the same package as the code that uses it, so it can definitely be package-private. Generally, returning collections is a tricky thing because Java collections are by definition mutable. We still do it, but if it can be done safer by restricting visibility (to package private in this case) or by only retrieving one keyed value at a time, it's generally worth it.